### PR TITLE
Func ref in array

### PR DESCRIPTION
**Pull request by mio991**
Hey, 
I changed methodes in Array from `(Object, StringName)` to `(FuncRef)` as talked about in #20545.

There are sadly still some const_casts in here because `ClassDB::bind_vararg_method()` doesn't support `const` methodes. Maybe I will work this out in the future but currently I don't know how.

### CODE REVIEW

File: core/array.cpp [hunk 1]
```diff
@@ -242,26 +243,24 @@ Array &Array::sort() {
...
-Array &Array::sort_custom(Object *p_obj, const StringName &p_function) {
+Array &Array::sort_custom(const FuncRef p_ref) { //TODO: FuncRef sort
 
-	ERR_FAIL_NULL_V(p_obj, *this);
+	//ERR_FAIL_NULL_V(p_ref, *this);
 
 	SortArray<Variant, _ArrayVariantSortCustom> avs;
-	avs.compare.obj = p_obj;
-	avs.compare.func = p_function;
+	avs.compare.ref = const_cast<FuncRef *>(&p_ref);
```
Comment (mio991) [inline core/array.cpp]: a sad const_cast

File: core/array.cpp [hunk 2]
```diff
@@ -312,13 +311,12 @@ int Array::bsearch(const Variant &p_value, bool p_before) {
...
-int Array::bsearch_custom(const Variant &p_value, Object *p_obj, const StringName &p_function, bool p_before) {
+int Array::bsearch_custom(const Variant &p_value, const FuncRef p_ref, bool p_before) { //TODO: FuncRef bsearch
 
-	ERR_FAIL_NULL_V(p_obj, 0);
+	//ERR_FAIL_NULL_V(p_ref, 0);
 
 	_ArrayVariantSortCustom less;
-	less.obj = p_obj;
-	less.func = p_function;
+	less.ref = const_cast<FuncRef *>(&p_ref);
```
Comment (mio991) [inline core/array.cpp]: another one

File: core/array.cpp [hunk 3]
```diff
@@ -355,6 +353,50 @@ Variant Array::pop_front() {
...
+			ret.append(this->get(i));
+		}
+	}
+
+	return ret;
+}
+
+Array Array::map(const FuncRef p_ref, const Variant &p_args) { //TODO: FuncRef map
+	Array ret;
+	FuncRef *ref = const_cast<FuncRef *>(&p_ref);
```
Comment (mio991) [inline core/array.cpp]: the last one

File: core/array.cpp [hunk 4]
```diff
@@ -355,6 +353,50 @@ Variant Array::pop_front() {
 	return Variant();
 }
 
+Array Array::filter(const FuncRef p_ref, const Variant &p_args) { //TODO: FuncRef filter
+	Array ret;
+	FuncRef *ref = const_cast<FuncRef *>(&p_ref);
```
Comment (mio991) [inline core/array.cpp]: and another one

**Comment by akien-mga** [comment]
The commit history is quite noisy with 11 commits, most of which seem to come from #20545 and should likely not be there.

**Comment by akien-mga** [comment]
Closing as this PR is not in a state that can be reviewed nor merged, as mentioned above.

**Pull request closed without merging**